apps/admin/backend/src/cvr_files.ts

```diff
@@ -0,0 +1,108 @@
+export interface CastVoteRecordFileMetadata {
+  readonly name: string;
+  readonly path: string;
+  readonly cvrCount: number;
+  readonly scannerIds: readonly string[];
+  readonly exportTimestamp: Date;
+  readonly isTestModeResults: boolean;
+}
+
+/**
+ * Gets the metadata, including the path, of cast vote record files found in
+ * the default location on a mounted USB drive. If there is no mounted USB
+ * drive or the USB drive appears corrupted then it will return an empty array.
+ */
+export async function listCastVoteRecordFilesOnUsb(
+  electionDefinition: ElectionDefinition,
+  logger: Logger
+): Promise<CastVoteRecordFileMetadata[]> {
+  const { election, electionHash } = electionDefinition;
+  const fileSearchResult = await listDirectoryOnUsbDrive(
+    join(
+      SCANNER_RESULTS_FOLDER,
+      generateElectionBasedSubfolderName(election, electionHash)
+    )
+  );
+
+  if (fileSearchResult.isErr()) {
+    const errorType = fileSearchResult.err().type;
+    switch (errorType) {
+      case 'no-entity':
+        await logger.log(LogEventId.CvrFilesReadFromUsb, 'system', {
+          message:
+            'No cast vote record files automatically found on USB drive. User is allowed to manually select files.',
+          disposition: 'success',
+        });
+        break;
+      case 'not-directory':
+      case 'permission-denied':
+        await logger.log(LogEventId.CvrFilesReadFromUsb, 'system', {
+          message:
+            'Error accessing cast vote record files on USB drive, which may be corrupted.',
+          disposition: 'failure',
+        });
+        break;
+      case 'no-usb-drive':
+      case 'usb-drive-not-mounted':
+        // we're just polling without a USB drive in these cases, no issue
+        break;
+      default:
+        throwIllegalValue(errorType);
+    }
+
+    return [];
+  }
+
+  const castVoteRecordFileMetadataList: CastVoteRecordFileMetadata[] = [];
+
+  for (const entry of fileSearchResult.ok()) {
+    if (
+      entry.type === FileSystemEntryType.File &&
+      entry.name.endsWith('.jsonl')
+    ) {
+      try {
+        const parsedFileInfo = parseCvrFileInfoFromFilename(entry.name);
+        if (parsedFileInfo) {
+          castVoteRecordFileMetadataList.push({
+            exportTimestamp: parsedFileInfo.timestamp,
+            cvrCount: parsedFileInfo.numberOfBallots,
+            isTestModeResults: parsedFileInfo.isTestModeResults,
+            name: entry.name,
+            path: entry.path,
+            scannerIds: [parsedFileInfo.machineId],
+          });
+        }
+      } catch (error) {
+        // The filename isn't able to be parsed as a valid CVR filename. We are
+        // only interested in valid CVR files so ignore it.
+        continue;
+      }
+    }
+  }
+
+  await logger.log(LogEventId.CvrFilesReadFromUsb, 'system', {
+    message: `Found ${castVoteRecordFileMetadataList.length} CVR files on USB drive, user shown option to load.`,
+    disposition: 'success',
+  });
```

Comment:
I'm searching for a pattern where we don't need to interleave log statements in our business logic, because it makes it harder to read and reason about the logic.

Any ideas? Maybe we could return a Result and then log success/failure one level up?

Reply:
No ideas, although your `Result` one is the best (maybe only) one I've heard. The current logging is pretty hard to read. I can imagine that more of our methods return `Result` objects and then we have a `logging.ts` that just has logging functions for each type of result. I am not wholly convinced that the current pattern is unreadable enough to make that worth it - the colocation also seems valuable.